conda_forge_tick/migrators/core.py

```diff
@@ -0,0 +1,861 @@
+        return feedstock_ctx.package_name + " v" + feedstock_ctx.attrs["new_version"]
+
+    def remote_branch(self, feedstock_ctx: FeedstockContext) -> str:
+        assert isinstance(feedstock_ctx.attrs["new_version"], str)
+        return feedstock_ctx.attrs["new_version"]
+
+    def migrator_uid(self, attrs: "AttrsTypedDict") -> "MigrationUidTypedDict":
+        n = super().migrator_uid(attrs)
+        assert isinstance(attrs["new_version"], str)
+        n["version"] = attrs["new_version"]
+        return n
+
+    def _extract_version_from_muid(self, h: dict) -> str:
+        return h.get("version", "0.0.0")
+
+    @classmethod
+    def new_build_number(cls, old_build_number: int) -> int:
+        return 0
+
+    def order(
+        self, graph: nx.DiGraph, total_graph: nx.DiGraph,
+    ) -> Sequence["PackageName"]:
+        return sorted(
+            graph, key=lambda x: (len(nx.descendants(total_graph, x)), x), reverse=True
+        )
+
+
+class GraphMigrator(Migrator):
+    def __init__(
+        self,
+        *,
+        name: Optional[str] = None,
+        graph: nx.DiGraph = None,
+        pr_limit: int = 0,
+        obj_version: Optional[int] = None,
+        piggy_back_migrations: Optional[Sequence[MiniMigrator]] = None,
+    ):
+        super().__init__(pr_limit, obj_version, piggy_back_migrations)
+        # TODO: Grab the graph from the migrator ctx
+        if graph is None:
+            self.graph = nx.DiGraph()
+        else:
+            self.graph = graph
+        self.name = name
+
+    def predecessors_not_yet_built(self, attrs: "AttrsTypedDict") -> bool:
+        # Check if all upstreams have been built
+        for node in self.graph.predecessors(attrs["feedstock_name"]):
+            payload = self.graph.nodes[node]["payload"]
+            muid = frozen_to_json_friendly(self.migrator_uid(payload))
+            if muid not in _sanitized_muids(
+                payload.get("PRed", [])
+            ) and not payload.get("archived", False,):
+                return True
+            # This is due to some PRed_json loss due to bad graph deploy outage
+            for m_pred_json in payload.get("PRed", []):
+                if m_pred_json["data"] == muid["data"]:
+                    break
+            else:
+                m_pred_json = None
+            # note that if the bot is missing the PR we assume it is open
+            # so that errors halt the migration and can be fixed
+            if (
+                m_pred_json
+                and m_pred_json.get("PR", {"state": "open"}).get("state", "") == "open"
+            ):
+                return True
+        return False
+
+
+class Replacement(Migrator):
+    """Migrator for replacing one package with another.
+
+    Parameters
+    ----------
+    old_pkg : str
+        The package to be replaced.
+    new_pkg : str
+        The package to replace the `old_pkg`.
+    rationale : str
+        The reason the for the migration. Should be a full statement.
+    """
```

Comment:
Doc string is out of date.

Reply:
Looks pretty up to date to me?

Reply:
Types are wrong but idk if the python typing is supposed to be in the doc strings too I guess.